src/lib/quoteEngine.js

```diff
@@ -43,15 +48,101 @@ const calculatePremiumPerYear = (coverAmount, basePrice, initialCapacityUsed, to
   const skipSurgePremium = amountOnSurgeSkip.mul(amountOnSurgeSkip).mul(SURGE_PRICE_RATIO).div(totalCapacity).div(2);
   const surgePremium = totalSurgePremium.sub(skipSurgePremium);
 
-  console.log('Cover amount   :', formatEther(coverAmount), 'nxm');
-  console.log('Amount on surge:', formatEther(amountOnSurge), 'nxm');
+  return basePremium.add(surgePremium);
+};
 
-  console.log('Base price     :', formatUnits(basePrice, 4), 'nxm');
-  console.log('Base premium   :', formatEther(basePremium), 'nxm');
-  console.log('Surge skipped  :', formatEther(skipSurgePremium), 'nxm');
-  console.log('Surge premium  :', formatEther(surgePremium), 'nxm');
+/**
+ * This function allocates each unit to the cheapest opportunity available for that unit
+ * at that time given the allocations at the previous points.
+ *
+ * To solve the allocation problem we split the amount A in U = `A / UNIT_SIZE + 1`
+ * units and use that for the rest of the computations.
+ *
+ * To allocate the units U we take each unit at a time from 0 to U,
+ * and allocate each unit to the *best* pool available in terms of price
+ * as long as the pool has enough capacity for that unit.
+ *
+ * If there is no pool with capacity available for a unit of allocation the function returns
+ * with an empty list (aborts allocation completely - no partial allocations supported).
+ *
+ * UNIT_SIZE is dynamically computed as 0.1% (1 / UNIT_DIVISOR) of the cover amount.
+ * It has a minimum which is 1e18 (can't go below 1 ETH or 1 DAI).
+ *
+ * Complexity O(n * p)  where n is the number of units in the amount and p is th  e number of pools
+ * @param coverAmount
+ * @param pools
+ * @param useFixedPrice
+ * @returns {{lowestCostAllocation: *, lowestCost: *}}
+ */
+const calculateOptimalPoolAllocation = (coverAmount, pools, useFixedPrice) => {
+
+  // set UNIT_SIZE to be a minimum of 1.
+  const UNIT_SIZE = coverAmount.div(UNIT_DIVISOR).gt(MIN_UNIT_SIZE) ? coverAmount.div(UNIT_DIVISOR) : MIN_UNIT_SIZE;
+
+  // compute the extra amount of units (0 or 1) to be added based on the division remainder.
+  const extra = coverAmount.mod(UNIT_SIZE).gt(0) ? 1 : 0;
+
+  const amountInUnits = coverAmount.div(UNIT_SIZE).add(extra).toNumber();
+
+  // the amount padding is the amount added artificially added when adding a whole unit
+  // to account for the remainder
+  const amountPadding = extra === 1 ? UNIT_SIZE.sub(coverAmount.mod(UNIT_SIZE)) : BigNumber.from(0);
```

Comment:
this doesn't look elegant at all. why do we need it? how can we implement this in a more elegant / cleaner way?

Reply:
if you split the input amount in fixed size chunks to run the algorithm on there's a remainder that is not exactly equal to  one chunk.

The remainder still needs to be allocated somewhere so the way i solve this is I map it to one full chunk.

To me this is the logical solution of how to solve this issue, maybe the naming can be improved?

Reply:
eliminated using @shark0der  suggestion of substracting instead of running a for loop from 0 to N